yt_dlp/extractor/crunchyroll.py

```diff
@@ -47,52 +49,52 @@ def _perform_login(self, username, password):
-                grant_type, key = 'etp_rt_cookie', 'accountAuthClientId'
-            else:
-                grant_type, key = 'client_id', 'anonClientId'
-
-            initial_state, app_config = self._get_embedded_json(self._download_webpage(
-                f'https://www.crunchyroll.com/{lang}', None, note='Retrieving main page'), None)
-            api_domain = app_config['cxApiParams']['apiDomain'].replace('beta.crunchyroll.com', 'www.crunchyroll.com')
-
-            auth_response = self._download_json(
-                f'{api_domain}/auth/v1/token', None, note=f'Authenticating with grant_type={grant_type}',
-                headers={
-                    'Authorization': 'Basic ' + str(base64.b64encode(('%s:' % app_config['cxApiParams'][key]).encode('ascii')), 'ascii')
-                }, data=f'grant_type={grant_type}'.encode('ascii'))
-            policy_response = self._download_json(
-                f'{api_domain}/index/v2', None, note='Retrieving signed policy',
-                headers={
-                    'Authorization': auth_response['token_type'] + ' ' + auth_response['access_token']
-                })
-            cms = policy_response.get('cms_web')
-            bucket = cms['bucket']
-            params = {
-                'Policy': cms['policy'],
-                'Signature': cms['signature'],
-                'Key-Pair-Id': cms['key_pair_id']
-            }
-            locale = traverse_obj(initial_state, ('localization', 'locale'))
-            if locale:
-                params['locale'] = locale
-            CrunchyrollBaseIE.params = (api_domain, bucket, params)
-        return CrunchyrollBaseIE.params
+    def _update_auth(self, lang):
+        if not self._QUERY:
+            self._QUERY = {}
+
+        if lang in self._QUERY:
+            return
+
+        webpage = self._download_webpage(
+            f'{self._BASE_URL}/{lang}', None, note=f'Retrieving main page (lang={lang or None})')
+
+        initial_state = self._search_json(r'__INITIAL_STATE__\s*=', webpage, 'initial state', None)
+        self._QUERY[lang] = traverse_obj(initial_state, {
+            'locale': ('localization', 'locale'),
+        }) or None
+
+        if self._AUTH_HEADERS:
+            return
+
+        if self.is_logged_in:
+            grant_type, key = 'etp_rt_cookie', 'accountAuthClientId'
+        else:
+            grant_type, key = 'client_id', 'anonClientId'
+
+        app_config = self._search_json(r'__APP_CONFIG__\s*=', webpage, 'app config', None)
+        auth_response = self._download_json(
+            f'{self._BASE_URL}/auth/v1/token', None, note=f'Authenticating with grant_type={grant_type}',
+            headers={
+                'Authorization': 'Basic ' + base64.b64encode(('%s:' % app_config['cxApiParams'][key]).encode()).decode()
+            }, data=f'grant_type={grant_type}'.encode())
+
+        self._AUTH_HEADERS = {'Authorization': auth_response['token_type'] + ' ' + auth_response['access_token']}
```

Comment:
This token expires in only 5 minutes. We need to take note of the timeout and re-authorize if it has expired. We shouldn't need to implement the special renew system the web client uses, we can just re-authorize from scratch, every new tab re-authorizes anyway, so this isn't abnormal behavior and shouldn't cause issues.

It might be a good idea to cache the `cxApiParams` so make this reauthorization work in a single request, though.

Reply:
It's definitely safe to cache. It might even be safe to hardcode in yt-dlp, I just didn't want to introduce another way for it to potentially break. I'm not sure it's ever actually changed yet.